pool/connection_pool.go

```diff
@@ -306,6 +318,9 @@ func (p *ConnectionPool) Close() []error {
 		}
 		p.addrsMutex.RUnlock()
 	}
+	for _, cancel := range p.ctxCancels {
+		cancel()
+	}
```

Comment:
The controller context and `endpoint.closed` channel is used for the same purpose. It controls a lifetime of a connection in the pool.

So you could replace `endpoint.closed` with the `cancel` object function and don't create a separate container for cancel functions + use the context object instead of the `endpoint.closed` channel.

Reply:
If we want to replace `close` channel - yes, it is.

Reply:
Updated. Now endpoint stores his context and cancel function.